zerver/lib/test_runner.py

```diff
@@ -101,69 +105,130 @@ def run_test(test):
+
+            msg = ("Import unexpectedly succeeded! Something is wrong. Try "
+                   "running `import %s` inside `./manage.py shell`.\n"
+                   "If that works, you may have introduced an import "
+                   "cycle.") % (actual_test_name,)
+            import_error = (Exception, Exception(msg), None)  # type: Tuple[Any, Any, Any]
+            result.addError(test, import_error)
             return True
         else:
-            print("Test doesn't have _pre_setup; something is wrong.")
-            print("Here's a debugger. Good luck!")
-            import pdb
-            pdb.set_trace()
+            msg = "Test doesn't have _pre_setup; something is wrong."
+            error_pre_setup = (Exception, Exception(msg), None)  # type: Tuple[Any, Any, Any]
+            result.addError(test, error_pre_setup)
+            return True
     test._pre_setup()
 
     start_time = time.time()
 
-    test.setUp()
-    try:
-        test_method()
-    except unittest.SkipTest as e:
-        print('Skipped:', e)
-    except Exception:
-        failed = True
-        traceback.print_exc()
-
-    test.tearDown()
+    test(result)  # unittest will handle skipping, error, failure and success.
 
     delay = time.time() - start_time
-    enforce_timely_test_completion(test_method, test_name, delay)
+    enforce_timely_test_completion(test_method, test_name, delay, result)
     slowness_reason = getattr(test_method, 'slowness_reason', '')
     TEST_TIMINGS.append((delay, test_name, slowness_reason))
 
     test._post_teardown()
     return failed
 
+class TextTestResult(runner.TextTestResult):
+    """
+    This class has unpythonic function names because base class follows
+    this style.
+    """
+    def addInfo(self, test, msg):
+        # type: (TestCase, Text) -> None
+        self.stream.write(msg)
+        self.stream.flush()
+
+    def addInstrumentation(self, test, data):
+        # type: (TestCase, Dict[str, Any]) -> None
+        append_instrumentation_data(data)
+
+    def startTest(self, test):
+        # type: (TestCase) -> None
+        TestResult.startTest(self, test)
+        self.stream.writeln("Running {}".format(full_test_name(test)))
+        self.stream.flush()
+
+    def addSuccess(self, *args, **kwargs):
+        # type: (*Any, **Any) -> None
+        TestResult.addSuccess(self, *args, **kwargs)
+
+    def addError(self, *args, **kwargs):
+        # type: (*Any, **Any) -> None
+        TestResult.addError(self, *args, **kwargs)
+
+    def addFailure(self, *args, **kwargs):
+        # type: (*Any, **Any) -> None
+        TestResult.addFailure(self, *args, **kwargs)
+
+    def addSkip(self, test, reason):
+        # type: (TestCase, Text) -> None
+        TestResult.addSkip(self, test, reason)
+        self.stream.writeln("** Skipping {}: {}".format(full_test_name(test),
+                                                        reason))
+        self.stream.flush()
```

Comment:
Yeah, we need to override `addSuccess`, `addError` and `addFailure` because we want to call the respective functions of `TestResult` and not `TextTestResult`. All the statistics are updated in `TestResult` while `TextTestResult` prints in the information on the console.